Pavel Trufanov HashMap

RailZz opened the pull request:
https://official.contest.yandex.ru/contest/7355/run-report/10700715/

RailZz committed:
Rename main.cpp to HashMap.cpp

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+#include <initializer_list>
+#include <stdexcept>
+#include <utility>
+#include <algorithm>
```
Инклуды неплохо бы отсортировать по алафвиту.

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+#include <utility>
+#include <algorithm>
+
+template<class KeyType, class ValueType, class Hash = std::hash<KeyType> > class HashMap {
```
class HashMap лучше на следующей строчке
C++11 позволяет писать >>

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+
+template<class KeyType, class ValueType, class Hash = std::hash<KeyType> > class HashMap {
+public:
+    HashMap(Hash hasher_ = Hash()):hasher(hasher_) {
```
Лучше привыкать писать Hash{}
Пробелы вокруг двоеточия

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+    HashMap(Hash hasher_ = Hash()):hasher(hasher_) {
+        setSize(HashMapSize);
+    }
+    template<class Iter>
```
Пустая строка

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+        }
+        setIterators();
+    }
+    HashMap(std::initializer_list<std::pair<const KeyType, ValueType>> init_list, Hash hasher_ = Hash()) :HashMap(init_list.begin(), init_list.end(), hasher_) {}
```
Слишком длинная строка, в идеале должно быть <80 символов

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+        return hasher;
+    }
+
+    void insert(const std::pair<const KeyType, ValueType>& add) {
```
Для std::pair<const KeyType, ValueType> стоит сделать typedef
Название параметра не должно быть глаголом, лучше value

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+    const int HashMapSize = 100000;
+    std::list<std::pair<const KeyType, ValueType>> data_;
+    std::vector<std::list<iterator>> remainers_;
+    Hash hasher;
```
Почему data_, но hasher? Лучше пусть подчёркивания будут у всего.

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+private:
+    const int HashMapSize = 100000;
+    std::list<std::pair<const KeyType, ValueType>> data_;
+    std::vector<std::list<iterator>> remainers_;
```
Может, remainders_? Непонятно, какой сысл у переменной.

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+    }
+
+private:
+    const int HashMapSize = 100000;
```
Константы лучше называть ВОТ_ТАК.

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    void print() const {
+        for (auto it = data_.begin(); it != data_.end(); ++it) {
```
for (const auto& it: data)

ifsmirnov commented on HashMap.cpp:
```diff
@@ -0,0 +1,148 @@
+    std::list<std::pair<const KeyType, ValueType>> data_;
+    std::vector<std::list<iterator>> remainers_;
+    Hash hasher;
+    void setSize(size_t size) {
```
Последний параметр конструктору не нужен. Можно заменить эту строчку на

`remainders_.assign(size, {});`

Название resize методу подходит лучше.

ifsmirnov reviewed (changes requested):
Это не всё, но для начала стоит это исправить.

Все функции стоит отделять пустыми строками.

Хештаблица должна уметь автоматически расширяться и поддерживать произвольный размер, а изначально создаваться маленького размера. Так, как у тебя -- нельзя.

RailZz committed:
Updated after review

https://official.contest.yandex.ru/contest/7355/run-report/10862459/

RailZz committed:
Update HashMap.cpp

RailZz committed:
Update HashMap.cpp

Pavel Trufanov
https://official.contest.yandex.ru/contest/7355/run-report/10865460/
